pkg/internal/vault/vault.go

```diff
@@ -144,7 +144,7 @@ func (v *Vault) Sign(csrPEM []byte, duration time.Duration) (cert []byte, ca []b
 
 	var caPem []byte = nil
 	if len(bundle.CAChain) > 0 {
-		caPem = []byte(bundle.CAChain[0])
+		caPem = []byte(strings.Join(bundle.CAChain, "\n"))
```

Comment:
I'm trying to find some docs on the actual `CAChain` field to ensure we don't accidentally make _another_ mistake here, but so far I can't find anything beyond: https://www.vaultproject.io/api-docs/secret/pki#sign-certificate

Do you know/have a reference to what clients should expect CAChain to contain?

Additionally, do we want the *chain* to be included here? Typically, the chain would be presented by the *server* (i.e. in `tls.crt` along with the leaf) and the *client* would trust the _root_ certificate instead.

Looking at the definition for `ToPEMBundle` here: https://github.com/hashicorp/vault/blob/84626b95f3195b48fa58ce5935aa25a4784465b6/sdk/helper/certutil/types.go#L137-L154 it appears that it already contains the full `CAChain`, so perhaps we need to query the `Read CA Certificate` endpoint instead? https://www.vaultproject.io/api-docs/secret/pki#read-ca-certificate

Reply:
When I started with tooling my intent was to match the output of the Vault PKI APIs; the `tls.crt`, `tls.key`, and `ca.crt` appeared to map to the `certificate`, `private_key`, and `ca_chain` fields, but the layout was different.

I am not finding great documentation on what "should be expected", so far I've noticed that:
- Prior to this PR (with only the first CA returned):
  - (client and server) certificates generated for etcd worked as expected with mTLS
  - client certificates did not work as expected with redis (`SSL_connect failed: certificate verify failed`)

With this PR, I am able to get certificates working with both services.

Reply:
That's what i'm talking about, there is a whole pattern when it comes to certificate management, there is the keypair (whatever.crt and whatever.key) and then there is the issuer (also known as CA) info on ca.crt (chain or not), and when one needs it creates the bundle to facilitate (result of `toPemBundle`)

@crandles changes will work for both cases, for the ones that uses `tls.crt` as a bundle or somehow use `ca.crt` to mount a local truststore.

> > It would surely bring challanges to solutions using only `tls.crt` as the bundle but it doesn't make the current approach right. Maybe we could provide another key to the secret: `bundle.pem` (tls.crt + ca.crt)
> 
> This looks similar: #843
To those who wish to have a combined PEM, this solution should be recomended ☝️

Reply:
AFAIK,

- `tls.crt` should have `certificate` + `ca_chain[1:len(ca_chain)]`
- `ca.crt` should have `ca_chain[len(ca_chain)]`

Explanation:

## What Vault API returns

Recap: When a certificate chain is given to [`/pki/intermediate/set-signed` Vault API](https://www.vaultproject.io/api-docs/secret/pki#set-signed-intermediate), then [`/pki/sign/:name`](https://www.vaultproject.io/api-docs/secret/pki#sign-certificate) returns the followings:

- `certificate`: a leaf certificate (that just issued)
- `issuing_ca`: a issuing CA certificate (when a Vault PKI mount is configured as an intermediate, then it's a intermediate CA)
- `ca_chain`: whole certificate chain, without a leaf certificate
  - `issuing_ca` + a chain given to `set-signed` API

## What server implementations expect

- `tls.crt` is expected to contain a leaf certificate _and_  intermediate CA certificate(s) but not to include a root CA
  - Most server implementations send additional certificates included in a PEM file to clients, as a certificate chain.
  - Technically, _the chain can include a root CA,_ but as clients are expected to have root CA in its trust store, so there's no need to transmit a redundant certificate to clients from a server.
  - (AFAIK, [SSL Labs test](https://www.ssllabs.com/ssltest/) would report warning when a server is sending redundant root CA certs)
- `ca.crt` is expected to contain a root CA certificate
  - A CA certificate which clients should trust in advance.

Reply:
I overlooked the hidden comments behind "load more" at this thread https://github.com/jetstack/cert-manager/pull/3340#discussion_r499078806  🙈 

The `ACME` issuer, which should be the most popular one, submits a `tls.crt` with a certificate chain, so I think it's safe to have a chain at Vault issuer as well.

Reply:
`Maybe we could provide another key to the secret: bundle.pem` this already is a feature request in our keyStores API, we should always follow `kubernetes.io/tls` for the issuers, the keystore controller will be able to provide more formats

Reply:
Generally speaking, it's meaningless or considered insecure to have a PEM bundle of certificate chain with intermediates, which may be revoked from its root CA afterwards. Clients shouldn't directly trust intermediates. Anyway, it can be up to users, but we shouldn't have it by default.

Reply:
Thanks everyone for the in-depth conversation and analysis! I agree with @sorah that the format should be:

`tls.crt` includes the leaf + chain but *not* the root cert
`ca.crt` includes _just_ the root certificate (not all issuer's are able to provide this)

In instances of a cross-signed certificate, I'd expect `ca.crt` to include *both* roots, though I don't think that's something we need to address today.

We should update the documentation to explicitly mention this format is expected too, and then work to ensure all our issuers keep to these rules.

---

Going back to this actual PR, it's not (yet) clear to me that this _is_ what we're doing (given Vault's own APIs/response formats too). How can we make this area of code clear, and clearly tested? Some comments on this area explaining the Vault response format would be brilliant, and then perhaps ensuring our test code is clear too (and maybe an explicit test for a function that converts a Vault response into a `{tls.key, tls.crt, ca.crt}` tuple, separate from the bulk of the logic here to make it easy/obvious to test and observe)

Reply:
Cross certificates should be put into `tls.crt`, not `ca.crt`. A root CA may have a cross certificate in addition to their self-signed root certificate. And from clients perspective, cross certs look like an intermediate between an existing trusted CA.